kube-runtime/src/controller/runner.rs

```diff
@@ -332,4 +373,82 @@ mod tests {
             Error::Readiness(delayed_init::InitDropped)
         ));
     }
+
+    // A Future that is Ready after the specified duration from its initialization.
+    struct DurationalFuture {
+        start: Instant,
+        ready_after: Duration,
+    }
+
+    impl DurationalFuture {
+        fn new(expires_in: Duration) -> Self {
+            let start = Instant::now();
+            DurationalFuture {
+                start,
+                ready_after: expires_in,
+            }
+        }
+    }
+
+    impl Future for DurationalFuture {
+        type Output = ();
+
+        fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+            let now = Instant::now();
+            if now.duration_since(self.start) > self.ready_after {
+                Poll::Ready(())
+            } else {
+                cx.waker().wake_by_ref();
+                Poll::Pending
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn runner_should_respect_max_concurrent_executions() {
+        pause();
+
+        let count = Arc::new(Mutex::new(0));
+        let (mut sched_tx, sched_rx) = mpsc::unbounded();
+        let mut runner = Box::pin(
+            Runner::new(scheduler(sched_rx), 2, |_| {
+                let mut num = count.lock().unwrap();
+                *num += 1;
+                DurationalFuture::new(Duration::from_secs(2))
+            })
+            .for_each(|_| async {}),
+        );
+
+        sched_tx
+            .send(ScheduleRequest {
+                message: 1,
+                run_at: Instant::now(),
+            })
+            .await
+            .unwrap();
+        assert!(poll!(runner.as_mut()).is_pending());
+        sched_tx
+            .send(ScheduleRequest {
+                message: 2,
+                run_at: Instant::now(),
+            })
+            .await
+            .unwrap();
+        assert!(poll!(runner.as_mut()).is_pending());
+        sched_tx
+            .send(ScheduleRequest {
+                message: 3,
+                run_at: Instant::now(),
+            })
+            .await
+            .unwrap();
+        assert!(poll!(runner.as_mut()).is_pending());
+        // Assert that we only ran two out of the three requests
+        assert_eq!(*count.lock().unwrap(), 2);
```

Comment:
Note to self; "num_pending_messages", "num_scheduled_messages", "num_running_messages" (and maybe a max if we set one) might make sense in the future to be exposed as optional metrics from the controller if that is ever a thing we want. It could indicate overload to users.